lib/services/publish-service.ts

```diff
@@ -0,0 +1,208 @@
+    private static PUBLIC_URL_DEFAULT_VALUE = "";
+    private static REMOVE_ALL_CONNECTIONS_MESSAGE = "All of them";
+
+    public allPublishConnections: IPublishConnection[];
+    public ftpPublishConnections: IPublishConnection[];
+
+    constructor(private $errors: IErrors,
+                private $fs: IFileSystem,
+                private $logger: ILogger,
+                private $progressIndicator: IProgressIndicator,
+                private $projectConstants: Project.IProjectConstants,
+                private $projectFilesManager: Project.IProjectFilesManager,
+                private $project: Project.IProject,
+                private $prompter: IPrompter,
+                private $server: Server.IServer) {
+
+        this.$project.ensureProject();
+
+        if (!this.$project.capabilities.publish) {
+            this.$errors.failWithoutHelp("The publish command is only applicable to projects of type %s", this.$projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.MobileWebsite);
+        }
+
+        this.readConnections().wait();
+        this.ftpPublishConnections = _.filter(this.allPublishConnections, c => c.type === PublishService.ALLOWED_CONNECTION_TYPE);
+    }
+
+    public publish(idOrUrl: string, username: string, password: string): IFuture<void> {
+        return (() => {
+            var projectName = this.$project.projectData.ProjectName;
+            var ftpConnectionData: Server.OfflineClientFtpConnectionData = {
+                RemoteUrl: idOrUrl,
+                ShouldPurge: options.force,
+                Username: username,
+                Password: password
+            };
+
+            if (!ftpConnectionData.Username) {
+                ftpConnectionData.Username = this.$prompter.getString("Username:").wait();
+            }
+
+            if (!ftpConnectionData.Password) {
+                ftpConnectionData.Password = this.$prompter.getPassword("Password:", {allowEmpty: true}).wait();
+            }
+
+            var id = parseInt(idOrUrl) - 1;
+
+            // the check whether it contains a dot is performed
+            // in case the user tries to publish to an IP address
+            if (!isNaN(id) && idOrUrl.indexOf('.') < 0) {
+                this.validateId(id);
+
+                ftpConnectionData.RemoteUrl = this.ftpPublishConnections[id].publishUrl;
+            }
+
+            var projectDir = this.$project.getProjectDir().wait();
+            this.$projectFilesManager.excludeFile(projectDir, PublishService.JSON_PUBLISH_FILE_NAME);
+            this.$project.importProject().wait();
+            this.$logger.printInfoMessageOnSameLine("Publishing");
+
+            try {
+                this.$progressIndicator.showProgressIndicator(this.$server.publish.publishToFtp(projectName, projectName, ftpConnectionData), 6000).wait();
+            } catch (err) {
+                this.$logger.printInfoMessageOnSameLine(os.EOL);
```

Comment:
This is a bug in typescript definitions :)

Try this code

```
			this.$logger.out("A");
			this.$logger.out();
			this.$logger.out("A");
```

tsc will complain that there's an error, but when you run the program, the expected empty line will be printed out.